cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -85,22 +129,52 @@ func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers []strin
 			return nil, fmt.Errorf("couldn't parse the kubeconfig file in the %s configmap: %v", bootstrapapi.ConfigMapClusterInfo, err)
 		}
 
-		fmt.Printf("[discovery] Cluster info signature and contents are valid, will use API Server %q\n", bootstrapConfig.Clusters[clusterName].Server)
+		// If no TLS root CA pinning was specified, we're done.
+		if pubKeyPins.Empty() {
+			fmt.Printf("[discovery] Cluster info signature and contents are valid and no TLS pinning was specified, will use API Server %q\n", endpoint)
+			return finalConfig, nil
+		}
+
+		// For each cluster, validate that its CA matches a pinned key and save that root into an x509.CertPool
+		pinnedRoots := x509.NewCertPool()
+		for _, cluster := range finalConfig.Clusters {
```

Comment:
Thanks. Instead of looping I added an assertion that there is only a single cluster.